parsl/executors/radical/executor.py

```diff
@@ -0,0 +1,388 @@
+"""RadicalPilotExecutor builds on the RADICAL-Pilot/Parsl
+"""
+import os
+import sys
+import time
+import parsl
+import queue
+import logging
+import inspect
+import typeguard
+import threading as mt
+
+import radical.pilot as rp
+import radical.utils as ru
+
+from functools import partial
+from typing import Optional, Union
+from concurrent.futures import Future
+
+from radical.pilot import PythonTask
+from parsl.utils import RepresentationMixin
+from parsl.executors.status_handling import NoStatusHandlingExecutor
+
+BASH = 'bash'
+PYTHON = 'python'
+
+logger = logging.getLogger(__name__)
+
+
+class RadicalPilotExecutor(NoStatusHandlingExecutor, RepresentationMixin):
+    """Executor designed for: executing heterogeneous tasks in terms of
+                              type/resource
+
+    The RadicalPilotExecutor system has the following components:
+
+      1. "start"    :creating the RADICAL-executor session and pilot.
+      2. "translate":unwrap/identify/ out of parsl task and construct RP task.
+      2. "submit"   :translating and submiting Parsl tasks the RADICAL-executor.
+      3. "shut_down":shutting down the RADICAL-executor components.
+
+    RADICAL Executor
+    ------------------------------------------------------------------------------------------------
+             Parsl DFK/dflow               |      Task Translator      |     RP-Client/Task-Manager
+    ---------------------------------------|---------------------------|----------------------------
+                                           |                           |
+    -> Dep. check ------> Parsl_tasks{} <--+--> Parsl Task/func/arg/kwg| tmgr.submit_Tasks(RP_tasks)
+     Data management          +dfk.submit  |             |             |
+                                           |             v             |
+                                           |     RP Task/Tasks desc. --+->
+    ------------------------------------------------------------------------------------------------
+    """
+
+    @typeguard.typechecked
+    def __init__(self,
+                 rpex_cfg,
+                 label: str = 'RPEX',
+                 bulk_mode: bool = False,
+                 resource: str = None,
+                 login_method: str = None,
+                 walltime: int = None,
+                 managed: bool = True,
+                 cores: Union[int, float] = float('inf'),
+                 gpus: Optional[int] = 0,
+                 worker_logdir_root: Optional[str] = ".",
+                 partition: Optional[str] = " ",
```

Comment:
The `Optional` annotation here means that I can pass in `None` as a value, distinct from the default value of " ".

Is that what you intend? If the value always has to be a string, never None, and " " is the right default value, remove the `Optional` annotation here.

Same for other lines nearby which use `Optional`.